feat: Bundle JSDoc-built TypeScript declaration file

brettz9 opened the pull request:
In the process of preparing ESLint for more type awareness, I figured we might avoid the need for an external `@types` package with the ability to maintain the types here (and to generate it out of JSDoc where possible as per https://github.com/eslint/espree/issues/529#issuecomment-1012561247 ).

One item I noticed in comparing with the declaration file in `@types/eslint-visitor-keys` is that `undefined` was allowed on properties, e.g.:

```ts
export interface VisitorKeys {
    readonly [type: string]: ReadonlyArray<string> | undefined;
}
```

In the PR, I have sought to preserve this ability, but I noticed that with the `unionWith` function, the additional keys argument couldn't include such properties as the current code would fail.

brettz9 commented:
I had assumed that the declaration file was building properly as tsc was not giving errors, but I see in testing it with a linked project, it is not.

```
espree.js:72:30 - error TS2306: File '/Users/<name>/eslint-visitor-keys/dist/eslint-visitor-keys.d.ts' is not a module.

72 import * as visitorKeys from "eslint-visitor-keys";
                                ~~~~~~~~~~~~~~~~~~~~~
```

This needs to be resolved before merging.

brettz9 commented:
The previous issue has now been resolved.

nzakas commented on .eslintrc.json:
```diff
@@ -7,6 +7,15 @@
+    "settings": {
+        "jsdoc": {
+            "mode": "typescript",
+            "preferredTypes": {
```
I think we should stick with uppercase Object, as we are doing that in most places already.

brettz9 replied:
Yeah, it's currently set as such in `eslint-config-eslint`.

But it gets a little complicated on what to do in this PR, however, and I need to cover some background, especially since the TypeScript doc details are kind of sparse and scattered and can't just refer you there.

Here's a summary of the situation between `object` vs. `Object`:

1. In non-JSDoc TypeScript, the two are different, with `Object` being [discouraged](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#number-string-boolean-symbol-and-object) and being  [treated like `any`](https://github.com/microsoft/TypeScript/issues/15105#issuecomment-350105361) (though testing in the playground, it appears to diverge from `any` in not allowing `null` or `undefined`), whereas `object` [allows](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#object-type) matching anything that is basically `typeof` "object" except for `null`. (Apparently `Object` works with whatever has `Object` on the prototype, thus allowing for strings, numbers, booleans, and symbols too, whereas `object` was [meant](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#object-type) specifically to allow for `Object.create(null)` type objects which do not have Object.)
1. In TypeScript-flavored JSDoc, there is meant to be no difference between the two per https://github.com/microsoft/TypeScript/issues/15105#issuecomment-350114644 (note also the response from the TypeScript team concurring), apparently because they were striving more for JSDoc compatibility than for consistency between TypeScript and TypeScript JSDoc
1. There is an issue ([#20555](https://github.com/microsoft/TypeScript/issues/20555)), reporting that with types like `object<string, number>`, `object` will get treated here as "any" and that issue aims to fix the unintentional differences within JSDoc between `object` and `Object` (even though it would mean less consistency between TypeScript and TypeScript JSDoc).

Based on this, it indeed shouldn't matter which we use in JSDoc in the future, and even now, `Object` may indeed seem preferable, as you suggest, until the issue is fixed, since it can be used in JSDoc.

However, when we automate the building of a declaration file, we get:

1. `Object` (as expected) occurs in the JSDoc (where it is at least not too strict to conflict with the TypeScript below if TypeScript gets preference (the TS playground at least seems to ignore the JSDoc), though it is too liberal in accepting some primitives) (in theory, `Object.create(null)` objects would not work, but the playground does not seem to enforce this)

1. A listing of `node: Object` in TypeScript proper, meaning in this context, that any item will be allowed including numbers and strings--something we don't want to allow here (and is worse than the existing `@types/eslint-visitor-keys` file in this regard).

```js

/**
 * Get visitor keys of a given node.
 * @param {Object} node The AST node to get keys.
 * @returns {readonly string[]} Visitor keys of the node.
 */
export function getKeys(node: Object): readonly string[];

```

By using "object", we instead get `object` in both places, thus avoiding the `node` argument being allowed to be too liberal ("object" ends up in the JSDoc too, meaning it currently would act like "any" there, but again, it appears the TypeScript overrides the JSDoc in TypeScript proper where it can occur.

(For what it's worth, in `eslint-plugin-jsdoc`, we are planning to enforce the behavior in `typescript` mode that by default `object` should be used in JSDoc for greater consistency with TypeScript though for `Object<string, number>` types, `Object` will be required due to the current TypeScript issue. So if you are persuaded, `eslint-config-eslint` could just rely on the default at that time, obviating the need for this extra config I added.)

nzakas replied:
Wow thanks for the detailed explanation. With all that background, I agree it seems like “object” is the better choice. 👍

nzakas commented on lib/index.js:
```diff
@@ -33,11 +33,11 @@ export function getKeys(node) {
  */
 export function unionWith(additionalKeys) {
-    const retv = Object.assign({}, KEYS);
+    const retv = /** @type {{ [type: string]: ReadonlyArray<string> }} */ (Object.assign({}, KEYS));
```
Can we name this type to avoid inlining long declarations like this?

brettz9 replied:
Sure, I've added a commit. Just note that this adds marginally to the declaration file, as adding a `@typedef` results in new `type`s being created (though that may actually be a convenience for users).

nzakas commented on tsconfig.json:
```diff
@@ -0,0 +1,19 @@
+    "declarationMap": true,
+    "emitDeclarationOnly": true,
+    "strict": true,
+    "target": "es5",
```
Should this be ES6?

brettz9 replied:
Although our Node `engine` versions should be safe for ES6, I don't know if you mean to preserve a wider browser compatibility? (To be more out-of-the-box browser ready, [eslint-plugin-ecmascript-compat](https://github.com/robatwilliams/es-compat/blob/master/packages/eslint-plugin-ecmascript-compat/README.md) could help, though I imagine the project may prefer to push that responsibility (and choice of supported environment) downstream.)

If just worrying about Node, if https://node.green/#ES2019 is any indication, we could go as high as `es2019`.

nzakas replied:
I think ES6 is well enough supported in browsers and Node.JS that we should target that as the minimum.

brettz9 replied:
I've added such a commit. We technically don't need any target here since we're not compiling JavaScript out of TypeScript, but I figured it might be good to settle on a target in case we do and wish to share the same basic config across projects.

nzakas commented on package.json:
```diff
@@ -4,9 +4,11 @@
   "exports": {
     ".": [
       {
+        "types": "./dist/index.d.ts",
```
Just checking: do we need this in two places?

brettz9 replied:
No, I've removed it. Per https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing , the other `types` is the fallback for older versions of TypeScript so probably best to keep that one.

Such `types` within `exports` might be usable if we were to export the `visitor-keys` file separately, as that has its own declaration file, but as we're not doing that, it's not necessary, so removed.

mdjermanovic commented on lib/visitor-keys.js:
```diff
@@ -1,3 +1,10 @@
+/**
+ * @typedef {{ readonly [type: string]: ReadonlyArray<string> }} KeysLooseReadonly
+ */
```
Why is this "Loose"? This definition looks the same as the definition for `KeysStrictReadonly` in the other file.

brettz9 replied:
Yes, you're right. This was a relic I had been kind of leaving in place in case there were concerns about our dropping the `| undefined` portion so we could go back to the Loose version at least here if desired (the discrepancy mentioned in the PR description) (`unionWith` couldn't use it with the code as currently factored). I've added a commit to update this (and if we don't need to permit `undefined` as `@types/eslint-visitor-keys` does, then I could also rename these types to stop calling them "Strict" too since this was only added to be in contrast to the "Loose" one, now having been dropped).

mdjermanovic replied:
Are there any observable differences in typescript checks when `| undefined` is added, or is it just redundant? I'd expect that it is implied that map-like objects cannot have all keys and that therefore `obj[key]` can be undefined.

brettz9 replied:
It gives an error with an explicit `undefined` (which is avoided when added to the type):

https://www.typescriptlang.org/play?#code/DYUwLgBAhgXBDeBYAUBNEBOIoBMD2AdsAJ4QDaYxADiHAM5gYCWBA5gLpwBK2+RxAQQwYoxADwNmbAHwoAvhAC8CFOmhwArgRwgAZixA55AbhRA

I

brettz9 replied:
Ok, sure. I just didn't know if that had been allowed for the sake of downstream libraries being able to compose objects with these properties removed or something. But I suppose that even if there would be some case for this, they could presumably manage this per their own project's needs. I can go ahead and change then.

mdjermanovic commented on package.json:
```diff
@@ -30,12 +33,14 @@
     "opener": "^1.5.2",
-    "rollup": "^2.52.1"
+    "rollup": "^2.52.1",
+    "typescript": "^4.6.0-dev.20220206"
```
Is it necessary for this feature to use this `-dev` version instead of a stable one?

brettz9 replied:
Apologies, left as a relic (along with `nodenext` resolution), both of which now reverted, which I believe I had been exploring due to problems with `outFile` and not finding the module issues.

(Using `outDir` turns out to have been the sole key for resolving that issue--with an apparently similar issue being reported, FWIW, by other users on [StackOverflow](https://stackoverflow.com/a/58534016/271577). That issue reports problems with module discovery when the output of `outFile` produces `declare module "index"`. Such output in this context is apparently what is responsible for preventing discovery of types by external projects (e.g., when I link my `espree` branch I'm working on to `eslint-visitor-keys`). I was only able to resolve this by use of `outDir`, which has the side effect of producing two declaration files instead of one).

mdjermanovic commented:
`index.d.ts` currently looks like this:

```ts
/**
 * Get visitor keys of a given node.
 * @param {object} node The AST node to get keys.
 * @returns {readonly string[]} Visitor keys of the node.
 */
export function getKeys(node: object): readonly string[];
/**
 * Make the union set with `KEYS` and given keys.
 * @param {KeysStrictReadonly} additionalKeys The additional keys.
 * @returns {KeysStrictReadonly} The union set.
 */
export function unionWith(additionalKeys: KeysStrictReadonly): KeysStrictReadonly;
export { KEYS };
export type KeysStrict = {
    [type: string]: readonly string[];
};
export type KeysStrictReadonly = {
    readonly [type: string]: readonly string[];
};
import KEYS from "./visitor-keys.js";
//# sourceMappingURL=index.d.ts.map
```

It exports some types, so someone might use them in their project. Should we aim for this to be stable (e.g., to avoid renaming exported types), and would it be possible to add some tests for `index.d.ts`?

brettz9 commented:
> <snip>... would it be possible to add some tests for `index.d.ts`?

FWIW, `@types/eslint-visitor-keys` has its declaration (from https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/eslint-visitor-keys/index.d.ts )

```js
export interface VisitorKeys {
    readonly [type: string]: ReadonlyArray<string> | undefined;
}

export const KEYS: VisitorKeys;
export function getKeys(node: {}): ReadonlyArray<string>;
export function unionWith(keys: VisitorKeys): VisitorKeys;
```

...and has some kind of pseudo-tests at https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/eslint-visitor-keys/eslint-visitor-keys-tests.ts .

TypeScript itself [recommends](https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html#tips) [`tsd`](https://github.com/SamVerschueren/tsd). For coverage though, tsd does not currently do so (there are some TypeScript test libraries which perform coverage, but I'm not sure that the libraries that do so would cover declaration files anyways--perhaps maybe only checking executable portions).

mdjermanovic commented on lib/index.js:
```diff
@@ -4,6 +4,11 @@
 import KEYS from "./visitor-keys.js";
 
+/**
+ * @typedef {{ [type: string]: ReadonlyArray<string> }} KeysStrict
+ * @typedef {{ readonly [type: string]: ReadonlyArray<string> }} KeysStrictReadonly
```
Thoughts about renaming `KeysStrict` -> `VisitorKeysWritable`, and `KeysStrictReadonly` -> `VisitorKeys`?

`VisitorKeys` would then be the same name as in [`@types/eslint-visitor-keys`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/eslint-visitor-keys/index.d.ts), which might be helpful for packages that are using `eslint-visitor-keys` with types, such as [`vue-eslint-parser`](https://github.com/vuejs/vue-eslint-parser/blob/f7a7a1ee2e4d14ff7062a053301779b5ec3415da/src/ast/traverse.ts#L6).

mdjermanovic commented on package.json:
```diff
@@ -30,12 +33,14 @@
   },
   "scripts": {
-    "prepare": "npm run build",
+    "prepare": "npm run build && npm run tsc",
```
I think `&& npm run tsc` should be in the `build` script.

brettz9 replied:
Sure, done...

brettz9 commented:
I looked into preparing tests, but ran into [an issue](https://github.com/SamVerschueren/tsd/issues/143) with the `tsd` testing framework, so unless you want plain TypeScript as the "tests", I don't know if you want any such testing in a separate PR.

brettz9 commented:
> > I looked into preparing tests, but ran into [an issue](https://github.com/SamVerschueren/tsd/issues/143) with the `tsd` testing framework, so unless you want plain TypeScript as the "tests", I don't know if you want any such testing in a separate PR.
> 
> I think it isn't critical to test `readonly`. If we could make tests that confirm that correct use of the APIs of this library does not cause typescript errors, that seems good enough for the start.

Ok, I've added some simple tests.

I also added a commit on top of that to rename per your type renaming suggestions.

mdjermanovic commented on test-d/index.test-d.ts:
```diff
@@ -0,0 +1,69 @@
+import { expectType, expectAssignable, expectError } from 'tsd';
+
+import { KEYS, getKeys, unionWith, VisitorKeysWritable, VisitorKeys } from "../lib/index.js";
```
Is this testing `dist/*.d.ts` files or source files?

I tried to make tests fail by editing `dist/index.d.ts` (I changed the return type of `getKeys()` to `number[]`), but `npm run tsd` didn't complain, so it was probably using the source files directly.

However, if I change this import path to `../`:

```suggestion
import { KEYS, getKeys, unionWith, VisitorKeysWritable, VisitorKeys } from "../";
```

then, it seems to be using `dist/*.d.ts` files, because `npm run tsd` now fails with `33:30 Argument of type readonly number[] is not assignable to parameter of type readonly string[].`

Would it make sense to change this path so that the tests use generated definitions, or it doesn't matter?

mdjermanovic commented on lib/index.js:
```diff
@@ -4,6 +4,11 @@
 import KEYS from "./visitor-keys.js";
 
+/**
+ * @typedef {{ [type: string]: ReadonlyArray<string> }} VisitorKeysWritable
```
Can we somehow _not_ export `VisitorKeysWritable`? It's only used internally, so it might be confusing why are we exporting a type that isn't used in the API.

mdjermanovic replied:
This issue might be related, but it seems there's no answer yet: https://github.com/microsoft/TypeScript/issues/46011

brettz9 replied:
Yes we can (at the small cost of our unit tests needing to redefine the type which is fine).

@nzakas had asked about using `@typedef` for  shorter type usage throughout the document, but yes, it seems like due to [this issue](https://github.com/microsoft/TypeScript/issues/46011) indicating we cannot control typedef exporting, using a typedef at this time requires us to make the type exported. I would think therefore that it a good idea to avoid using typedefs for local types at this time.

We could at least use newlines to make the type more readable when inline, e.g.,

```js
const retv = /** @type {{
    [type: string]: ReadonlyArray<string>
}} */ (Object.assign({}, KEYS));
```

mdjermanovic replied:
Yes, if there's currently no way to tell that some types are internal, I think it's better to inline types we don't want to export until that issue gets resolved.

brettz9 replied:
Ok, I've added a commit to avoid exporting that type. I did redefine the type within the testing file for convenience and clarity, but that testing file isn't being exported (and the type isn't defined that way anyways).

mdjermanovic commented on test-d/index.test-d.ts:
```diff
@@ -0,0 +1,71 @@
+
+import { KEYS, getKeys, unionWith, VisitorKeys } from "../";
+
+type VisitorKeysWritable = { [type: string]: ReadonlyArray<string> };
```
Why do we need this in the tests?

brettz9 replied:
Was just for convenience--like defining a `const` in a Mocha test. There's no `export` on this `type`, so it seems to me to fit with the testing approach here.

mdjermanovic replied:
No problem with that, but what are we testing with `VisitorKeysWritable`?

brettz9 replied:
I see. Yes, the tests had indeed only been relevant before because we were actually exporting the type. Now removed.

mdjermanovic reviewed (approved):
LGTM, thanks!

mdjermanovic commented:
It looks like all comments by @nzakas have been addressed, so I'm merging this now for the release.

The two new files we'll publish now have the following content.

`dist/index.d.ts`:

```ts
/**
 * Get visitor keys of a given node.
 * @param {object} node The AST node to get keys.
 * @returns {readonly string[]} Visitor keys of the node.
 */
export function getKeys(node: object): readonly string[];
/**
 * Make the union set with `KEYS` and given keys.
 * @param {VisitorKeys} additionalKeys The additional keys.
 * @returns {VisitorKeys} The union set.
 */
export function unionWith(additionalKeys: VisitorKeys): VisitorKeys;
export { KEYS };
export type VisitorKeys = {
    readonly [type: string]: readonly string[];
};
import KEYS from "./visitor-keys.js";
//# sourceMappingURL=index.d.ts.map
```

`dist/visitor-keys.d.ts`:

```ts
export default KEYS;
export type VisitorKeys = import('./index.js').VisitorKeys;
/**
 * @typedef {import('./index.js').VisitorKeys} VisitorKeys
 */
/**
 * @type {VisitorKeys}
 */
declare const KEYS: VisitorKeys;
//# sourceMappingURL=visitor-keys.d.ts.map
```

I verified that [DefinitelyTyped/types/eslint-visitor-keys/eslint-visitor-keys-tests.ts](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/eslint-visitor-keys/eslint-visitor-keys-tests.ts) works well with `eslint-visitor-keys` package installed from this branch, so it's unlikely that this change will break anyone's `tsc` build. I've also verified that these declarations prevent invalid use of eslint-visitor-keys APIs in a TS code.
